README.md
# Overview

Liquidation of assets environment.

This project was generated with [cookiecutter](https://github.com/audreyr/cookiecutter) using [jacebrowning/template-python](https://github.com/jacebrowning/template-python).

[![Unix Build Status](https://img.shields.io/travis/juan-carlos-calvo/finance-rl.svg?label=unix)](https://travis-ci.org/juan-carlos-calvo/finance-rl)
[![Windows Build Status](https://img.shields.io/appveyor/ci/juan-carlos-calvo/finance-rl.svg?label=windows)](https://ci.appveyor.com/project/juan-carlos-calvo/finance-rl)
[![Coverage Status](https://img.shields.io/coveralls/juan-carlos-calvo/finance-rl.svg)](https://coveralls.io/r/juan-carlos-calvo/finance-rl)
[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/juan-carlos-calvo/finance-rl.svg)](https://scrutinizer-ci.com/g/juan-carlos-calvo/finance-rl)
[![PyPI Version](https://img.shields.io/pypi/v/finance-rl.svg)](https://pypi.org/project/finance-rl)
[![PyPI License](https://img.shields.io/pypi/l/finance-rl.svg)](https://pypi.org/project/finance-rl)

# Setup

## Requirements

* Python 3.8+

## Installation

Install it directly into an activated virtual environment:

```text
$ pip install finance-rl
```

or add it to your [Poetry](https://poetry.eustace.io/) project:

```text
$ poetry add finance-rl
```

# Usage

After installation, the package can imported:

```text
$ python
>>> import finance_rl
>>> finance_rl.__version__
```


CHANGELOG.rst

Changelog
=========

0.0.0 (2021-01-23)
------------------

* First release on PyPI.


tests/test_liquidation_gym.py

from liquidation_gym import main


def test_main():
    pass


tests/test_envs/test_models.py
import hypothesis.strategies as st
import numpy as np
from hypothesis import given

from finance_rl.envs import AlmgrenChrissParams


@given(
    lt=st.integers(1, 360),
    nt=st.integers(1, 10000),
    etap=st.floats(1e-5, 0.01, exclude_min=True),
    gamma=st.floats(1e-5, 0.01, exclude_min=True),
    llambda=st.floats(0, 1e-5, exclude_min=True),
    ssv=st.floats(1, 1000),
    xs=st.lists(
        st.floats(min_value=0, exclude_min=True, max_value=1e6), min_size=0, max_size=7
    ),
)
def test_info(lt, nt, etap, gamma, llambda, ssv, xs):
    etah = 0.5 * gamma * lt / nt + etap
    params = AlmgrenChrissParams(lt, nt, etah, gamma, llambda, ssv, *xs)
    assert params.tau == params.liquidation_time / params.num_n
    assert params.eta_hat == params.eta - (0.5 * params.gamma * params.tau)
    assert params.kappa_hat == np.sqrt(
        (params.llambda * params.singleStepVariance) / params.eta_hat
    )
    assert params.kappa == (
        np.arccosh((((params.kappa_hat ** 2) * (params.tau ** 2)) / 2) + 1) / params.tau
    )


tests/test_envs/test_liquidation_env.py
import gym
import hypothesis.strategies as st
import numpy as np
from hypothesis import assume, given

from finance_rl.envs import EnvParams
from finance_rl.envs.models import ETA, GAMMA


ENV_ID = 'liquidation-v0'


@given(
    liquidation_time=st.integers(1, 1000),
    num_n=st.integers(1, 1000),
    llambda=st.floats(0, 1e-5, exclude_min=True),
)
def test_reset(liquidation_time, num_n, llambda):
    assume(liquidation_time / num_n < 2 * ETA / GAMMA)
    env_params = EnvParams(
        liquidation_time=liquidation_time, num_n=num_n, llambda=llambda
    )
    env = gym.make(ENV_ID)
    env.reset(env_params=env_params)
    np.testing.assert_allclose(
        env.state.initial_state,
        np.array(
            list(env.state.logReturns)
            + [
                env.state.timeHorizon / env.alm_params.num_n,
                env.state.shares_remaining / env.alm_params.total_shares,
            ]
        ),
    )
    assert (
        env.transactions.prevPrice == env.alm_params.startingPrice
    ), 'prev price not correctly initialized'
    if env.alm_params.kappa > 0:
        breakpoint()
        assert env.transactions.prevUtility == env.compute_AC_utility(
            env.alm_params.total_shares
        ), 'prev utility not correctly initialized'


def test_AC_variance():
    actual = 204.3095586503
    env = gym.make(ENV_ID)
    result = env.get_AC_variance(30)
    np.testing.assert_almost_equal(actual, result)


def test_AC_expected_shortfall():
    actual = 1.87537369133
    env = gym.make(ENV_ID)
    result = env.get_AC_expected_shortfall(30)
    np.testing.assert_almost_equal(actual, result)


def test_AC_utility():
    actual = 1.87557800
    env = gym.make(ENV_ID)
    result = env.compute_AC_utility(30)
    np.testing.assert_almost_equal(actual, result)
    env_params = EnvParams(liquidation_time=0)
    env.reset(env_params=env_params)
    assert env.compute_AC_utility(0) == 0


scent.py
"""Configuration file for sniffer."""

import time
import subprocess

from sniffer.api import select_runnable, file_validator, runnable
try:
    from pync import Notifier
except ImportError:
    notify = None
else:
    notify = Notifier.notify


watch_paths = ["finance_rl", "tests"]


class Options:
    group = int(time.time())  # unique per run
    show_coverage = False
    rerun_args = None

    targets = [
        (('make', 'test-unit', 'DISABLE_COVERAGE=true'), "Unit Tests", True),
        (('make', 'test-all'), "Integration Tests", False),
        (('make', 'check'), "Static Analysis", True),
        (('make', 'docs'), None, True),
    ]


@select_runnable('run_targets')
@file_validator
def python_files(filename):
    return filename.endswith('.py') and '.py.' not in filename


@select_runnable('run_targets')
@file_validator
def html_files(filename):
    return filename.split('.')[-1] in ['html', 'css', 'js']


@runnable
def run_targets(*args):
    """Run targets for Python."""
    Options.show_coverage = 'coverage' in args

    count = 0
    for count, (command, title, retry) in enumerate(Options.targets, start=1):

        success = call(command, title, retry)
        if not success:
            message = "✅ " * (count - 1) + "❌"
            show_notification(message, title)

            return False

    message = "✅ " * count
    title = "All Targets"
    show_notification(message, title)
    show_coverage()

    return True


def call(command, title, retry):
    """Run a command-line program and display the result."""
    if Options.rerun_args:
        command, title, retry = Options.rerun_args
        Options.rerun_args = None
        success = call(command, title, retry)
        if not success:
            return False

    print("")
    print("$ %s" % ' '.join(command))
    failure = subprocess.call(command)

    if failure and retry:
        Options.rerun_args = command, title, retry

    return not failure


def show_notification(message, title):
    """Show a user notification."""
    if notify and title:
        notify(message, title=title, group=Options.group)


def show_coverage():
    """Launch the coverage report."""
    if Options.show_coverage:
        subprocess.call(['make', 'read-coverage'])

    Options.show_coverage = False


finance_rl/envs/liquidation_env.py
import gym
import numpy as np
from gym import error, spaces, utils
from gym.utils import seeding

from .models import EnvParams, Transactions


class LiquidationEnv(gym.Env):
    metadata = {'render.modes': ['human']}

    def __init__(self, env_params: EnvParams = EnvParams()):
        self.info = env_params.info
        self.state = env_params.state
        self.alm_params = env_params.alm_params
        self.fin_params = env_params.fin_params
        self.transactions = Transactions()

    def step(self, action):
        pass

    def reset(self, env_params: EnvParams = EnvParams()):
        self.__init__(env_params)
        self.state.initial_state = np.array(
            list(self.state.logReturns)
            + [
                self.state.timeHorizon / self.alm_params.num_n,
                self.state.shares_remaining / self.alm_params.total_shares,
            ]
        )
        self.transactions.prevPrice = self.alm_params.startingPrice
        self.transactions.prevUtility = self.compute_AC_utility(
            self.alm_params.total_shares
        )

    def render(self, mode='human', close=False):
        pass

    def get_AC_variance(self, sharesToSell: int) -> float:
        ft = 0.5 * (self.alm_params.singleStepVariance) * (sharesToSell ** 2)
        nst = self.alm_params.tau * np.sinh(
            self.alm_params.kappa * self.alm_params.liquidation_time
        ) * np.cosh(
            self.alm_params.kappa
            * (self.alm_params.liquidation_time - self.alm_params.tau)
        ) - self.alm_params.liquidation_time * np.sinh(
            self.alm_params.kappa * self.alm_params.tau
        )
        dst = (
            np.sinh(self.alm_params.kappa * self.alm_params.liquidation_time) ** 2
        ) * np.sinh(self.alm_params.kappa * self.alm_params.tau)
        st = nst / dst
        return ft * st

    def get_AC_expected_shortfall(self, sharesToSell: int) -> float:
        ft = 0.5 * self.alm_params.gamma * (sharesToSell ** 2)
        st = self.alm_params.epsilon * sharesToSell
        tt = self.alm_params.eta_hat * (sharesToSell ** 2)
        nft = np.tanh(0.5 * self.alm_params.kappa * self.alm_params.tau) * (
            self.alm_params.tau
            * np.sinh(2 * self.alm_params.kappa * self.alm_params.liquidation_time)
            + 2
            * self.alm_params.liquidation_time
            * np.sinh(self.alm_params.kappa * self.alm_params.tau)
        )
        dft = (
            2
            * (self.alm_params.tau ** 2)
            * (np.sinh(self.alm_params.kappa * self.alm_params.liquidation_time) ** 2)
        )
        fot = nft / dft
        return ft + st + (tt * fot)

    def compute_AC_utility(self, sharesToSell: int) -> float:
        # Calculate the AC Utility according to pg. 13 of the AC paper
        if self.alm_params.liquidation_time == 0:
            return 0
        E = self.get_AC_expected_shortfall(sharesToSell)
        V = self.get_AC_variance(sharesToSell)
        return E + self.alm_params.llambda * V


finance_rl/envs/models.py
import collections
from typing import Iterable

import numpy as np
from pydantic import BaseModel, root_validator, validator
from pydantic.dataclasses import dataclass


class Info(BaseModel):
    done: bool = False
    implementation_shortfall: float = None
    expected_shortfall: float = None
    expected_variance: float = None
    utility: float = None
    price: float = None
    share_to_sell_now: int = None
    currentPermanentImpact: float = None
    currentTemporaryImpact: float = None
    exec_price: float = None


ANNUAL_VOLAT = 0.12  # Annual volatility in stock price
BID_ASK_SP = 1 / 8  # Bid-ask spread
DAILY_TRADE_VOL = 5e6  # Average Daily trading volume
TRAD_DAYS = 250  # Number of trading days in a year
DAILY_VOLAT = ANNUAL_VOLAT / np.sqrt(TRAD_DAYS)  # Daily volatility in stock price


class FinantialParams(BaseModel):
    anv: float = ANNUAL_VOLAT
    basp: float = BID_ASK_SP
    dtv: int = DAILY_TRADE_VOL
    dpv: float = DAILY_VOLAT


TOTAL_SHARES = 1000000  # Total number of shares to sell
STARTING_PRICE = 50  # Starting price per share
LLAMBDA = 1e-6  # Trader's risk aversion
LIQUIDATION_TIME = 60  # How many days to sell all the shares.
NUM_N = 60  # Number of trades
EPSILON = BID_ASK_SP / 2  # Fixed Cost of Selling.
SINGLE_STEP_VARIANCE = (
    DAILY_VOLAT * STARTING_PRICE
) ** 2  # Calculate single step variance
ETA = BID_ASK_SP / (
    0.01 * DAILY_TRADE_VOL
)  # Price Impact for Each 1% of Daily Volume Traded
GAMMA = BID_ASK_SP / (0.1 * DAILY_TRADE_VOL)  # Permanent Impact Constant


@dataclass
class AlmgrenChrissParams:
    liquidation_time: int = LIQUIDATION_TIME
    num_n: int = NUM_N
    eta: float = ETA
    gamma: float = GAMMA
    llambda: float = LLAMBDA
    singleStepVariance: float = SINGLE_STEP_VARIANCE
    total_shares: int = TOTAL_SHARES
    startingPrice: int = STARTING_PRICE
    epsilon: float = EPSILON
    tau: float = None
    eta_hat: float = None
    kappa_hat: float = None
    kappa: float = None

    def __post_init_post_parse__(self):
        self.tau = self.liquidation_time / self.num_n
        self.eta_hat = self.eta - (0.5 * self.gamma * self.tau)
        self.kappa_hat = np.sqrt(
            (self.llambda * self.singleStepVariance) / self.eta_hat
        )
        self.kappa = (
            np.arccosh((((self.kappa_hat ** 2) * (self.tau ** 2)) / 2) + 1) / self.tau
            if self.tau > 0
            else 1
        )


class State(BaseModel):
    shares_remaining: int = TOTAL_SHARES
    timeHorizon: int = NUM_N
    logReturns: Iterable = collections.deque(np.zeros(6))
    prevImpactedPrice: int = STARTING_PRICE
    transacting: bool = False
    trade_number: int = 0
    initial_state: Iterable = None


class EnvParams(BaseModel):
    info: Info = Info()
    state: State = State()
    alm_params: AlmgrenChrissParams = AlmgrenChrissParams()
    fin_params: FinantialParams = FinantialParams()

    @root_validator(pre=True)
    def parse_init(cls, params):
        st = {'timeHorizon': params.get('num_n', NUM_N)}
        kwargs = {}
        kwargs['alm_params'] = params
        kwargs['state'] = st
        return kwargs


class Transactions(BaseModel):
    tolerance: int = 1
    totalCapture: int = 0
    prevPrice: int = None
    totalSSSQ: int = 0
    totalSRSQ: int = 0
    prevUtility: int = None

    # # Set the variables for the initial state
    # self.shares_remaining = self.total_shares
    # self.timeHorizon = self.num_n
    # self.logReturns = collections.deque(np.zeros(6))

    # # Set the initial impacted price to the starting price
    # self.prevImpactedPrice = self.startingPrice

    # # Set the initial transaction state to False
    # self.transacting = False

    # # Set a variable to keep trak of the trade number
    # self.k = 0
